src/python/pants/core/util_rules/system_binaries.py

```diff
@@ -320,6 +324,39 @@ class DiffBinary(BinaryPath):
     pass
 
 
+class GitBinaryException(Exception):
+    pass
+
+
+class GitBinary(BinaryPath):
+    def invoke(self, cmd: list[str]) -> str:
```

Comment:
Not blocking, but it would be good to add a docstring explaining that it is only safe for us to use subprocess because this type is marked as not cachable.

Speaking of, I am still a little confused which types need to have that mark in which don't. For example if this needs it?

Reply:
A type should be marked `EngineAwareReturnType.cacheable=False` if it should be re-computed in each `Session`: see https://github.com/pantsbuild/pants/blob/cc4e1800416dadbf631f06799f840ec3242dadc0/src/python/pants/engine/engine_aware.py#L66-L72

The strategy here is to use that facility to cause _consumers_ of git types to be forced to re-run: so if `GitWorktree` is marked `cacheable=False`, any `@rule` which consumes it should re-run in each session, iff `GitWorktree` has an `__eq__` implementation which will cause it to not be equal (I'll comment elsewhere with that fact).

While this type could in theory also be marked, I think that the expectation is that only the `GitWorktree` should be directly consumed, and that this type is an implementation detail.

----

But yea: this is definitely a fairly fragile way to expose these types: it is easy to get wrong. But it feels worthwhile to expose them internally first with this API and see how that goes.

Reply:
I've pushed a commit to clarify and test this.